sdk/go/pkg/state/client.go

```diff
@@ -0,0 +1,284 @@
+
+func (b *boundStateBuilder) WithEnvScope() ScopedState {
+	return b.WithCustomScope(StateScope{
+		OrganizationID: b.organizationID,
+		ProjectID:      &b.projectID,
+		EnvironmentID:  &b.environmentID,
+	})
+}
+
+func (b *boundStateBuilder) WithCustomScope(scope StateScope) ScopedState {
+	return &boundState{
+		httpClient:     b.httpClient,
+		baseURL:        b.baseURL,
+		token:          b.token,
+		organizationID: b.organizationID,
+		projectID:      b.projectID,
+		environmentID:  b.environmentID,
+		scope:          scope,
+	}
+}
+
+type ScopedState interface {
+	Get(ctx context.Context, key string) (json.RawMessage, uint32, error)
+	Set(ctx context.Context, key string, updateFn func(old json.RawMessage) (json.RawMessage, error)) error
+	SetKV(ctx context.Context, key string, value json.RawMessage, version uint32) error
+	Delete(ctx context.Context, key string) error
+}
+
+type boundState struct {
+	httpClient     *http.Client
+	baseURL        string
+	token          string
+	organizationID string
+	projectID      string
+	environmentID  string
+	scope          StateScope
+}
+
+type stateResponse struct {
+	Value   json.RawMessage `json:"value"`
+	Version uint32          `json:"version"`
+}
+
+func (b *boundState) injectHeaders(req *http.Request) {
+	req.Header.Set("X-Eaas-State-Token", b.token)
+	req.Header.Set("X-Organization-ID", b.organizationID)
+	req.Header.Set("X-Project-ID", b.projectID)
+	req.Header.Set("X-Environment-ID", b.environmentID)
+	req.Header.Set("Content-Type", "application/json")
+}
+
+func (b *boundState) Get(ctx context.Context, key string) (json.RawMessage, uint32, error) {
+	var zero json.RawMessage
+	value, version, err := b.getRaw(ctx, key)
+	if err != nil {
+		return zero, 0, sdk.NewErrNotFound(fmt.Sprintf("key not found: %s", key))
+	}
+	return value, version, nil
+}
+
+// Developer passes key and value to set
+func (b *boundState) SetKV(ctx context.Context, key string, value json.RawMessage, version uint32) error {
+
+	// Send update with version for OCC
+	body := map[string]any{
+		"scope":   b.scope,
+		"key":     key,
+		"value":   value,
+		"version": version,
+	}
+	buf, _ := json.Marshal(body)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPut, b.baseURL, bytes.NewReader(buf))
+	if err != nil {
+		return err
+	}
+	b.injectHeaders(req)
+
+	resp, err := b.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusConflict {
+		return sdk.NewErrConflict("version conflict on set")
+	}
+	if resp.StatusCode != http.StatusOK {
+		body, _ := io.ReadAll(resp.Body)
+		return sdk.NewErrFailed(fmt.Sprintf("set failed: %s", string(body)))
+	}
+
+	// Success
+	return nil
+}
+
+// Set abstracts create/update with OCC retry logic
+// Developer passes a transformation function that takes the old value and returns the new one.
+func (b *boundState) Set(ctx context.Context, key string, updateFn func(old json.RawMessage) (json.RawMessage, error)) error {
```

Comment:
Handles OCC with retries limited to 5, takes a transformation function ( which takes old state value ) and modifies the state field attributes. Could be useful for cases like manage number of nodes in state, or to identify diff and act upon it